public/app/plugins/datasource/grafana-azure-monitor-datasource/components/QueryEditor/QueryEditor.tsx

```diff
@@ -17,15 +17,34 @@ import ArgQueryEditor from '../ArgQueryEditor';
 import ApplicationInsightsEditor from '../ApplicationInsightsEditor';
 import InsightsAnalyticsEditor from '../InsightsAnalyticsEditor';
 import { Space } from '../Space';
+import { debounce } from 'lodash';
+import useDefaultQuery from './useDefaultQuery';
 
 export type AzureMonitorQueryEditorProps = QueryEditorProps<
   AzureMonitorDatasource,
   AzureMonitorQuery,
   AzureDataSourceJsonData
 >;
 
-const QueryEditor: React.FC<AzureMonitorQueryEditorProps> = ({ query, datasource, onChange }) => {
+const QueryEditor: React.FC<AzureMonitorQueryEditorProps> = ({
+  query: baseQuery,
+  datasource,
+  onChange,
+  onRunQuery: baseOnRunQuery,
+}) => {
   const [errorMessage, setError] = useLastError();
+  const onRunQuery = useMemo(() => debounce(baseOnRunQuery, 500), [baseOnRunQuery]);
+
+  const onQueryChange = useCallback(
+    (newQuery: AzureMonitorQuery) => {
+      onChange(newQuery);
+      onRunQuery();
```

Comment:
Does onRunQuery depend on there being a new query getting successfully set or is it separate side effect? If so I wonder if it should be put in a useEffect and run when query changes instead since I don't know that we can guarantee that onChange will be finished setting the new Query by the time onRunQuery is called.

Reply:
That's a great question! I'll do some research.

Reply:
we do seem to have this pattern everywhere including in some docs, so maybe it's fine?? I don't really understand why we don't run into race conditions but I also feel like maybe I can let it go haha.

Reply:
my guess is that `onChange` just isn't an async function - it updates a redux store synchronously, and then `onRunQuery` reads that store synchronously and dispatches a query (which is async)

Reply:
it's just weird because I assume onChange uses setState under the hood which is async (react like batch calls them i think, but maybe the order is always predictable? idk it's been a minute i guess since I dug into this!)